### DISCUSSION TITLE
My 2 cents on giving value

### DISCUSSION THREAD

**Discussion by Buzimu** [Godot-Community-Games-Org/game1#6] [Ideas]
What I would like to see from this community, this collection of minds, is two branches of things.

1. As per the original idea, a collection of little or large games both for fun and to serve as little templates or starting off points for people.

2. A collection of 'modules', 'plugins', where we can flesh out a single 'feature', where people can effectively drag and drop into their own game to enable that feature within their own game.
Examples from the top of my head:
- A module to attach to your player which enables an inventory
- A module to attach to your map which enables a day/night cycle
- A module to attach to your map which enables a dev console UI and outputs stats to the screen
- A module to attach to your enemies which enables it to drop an item from a loot table on death.

Let me know your thoughts on this

**Comment by ryankmims** [comment]
I like the idea of this game serving as a jumping off point, or a go-to point for learning 100%. While the Godot documents are fantastic, far beyond most documentation, there is a serious lack of explanation of simple things like Questing and Inventory systems for beginners. Even just a "map" system.

I don't know if the things you're talking about could be componentized and turned into modules that can just be added to your game. But if someone is curious like, "how do I make quests?" they will have a resource for learning that for sure if we can get this working

**Comment by Yonodactyl** [comment]
I like the two ideas you're hoping for with the project. I think if we are hoping to be a bastion of resources for the new developer that means we might want to have some solid principles and standards to what we commit. Documented code, references to the docs, proper use of threading, etc...

I said it earlier, but this almost feels like we are creating our own little "collective" of developers, like the [Sokpop](https://sokpop.co/) guys!

**Comment by GodotCommunityGames** [comment]
Yes! I love those ideas. And yea Yonodactyl I like that collective of developers idea.

I myself am still just getting started learning Godot, but I've been starting to make a couple general composable components in my projects that I can just add to characters to give them whatever traits. We could have a whole "Comoponents" repo where we try to just build composable scenes that can act as individual features in games to be added to other scenes. Stuff like basic features of characters, inventory system, minimap, whatever, with commented documentation at the top of each component script that describes the API (how to use the component).

In this Components repo perhaps we could also put other sort of concepts like an Signal Bus, an Audio Manager, a Scene Manager, etc.

All these things could be used in community games as well as for anyone to use in their own projects, either as drop-in components/modules or templates that they can then edit to their own needs. In effect this would build up a bunch of usable components but also essentially tutorial learning resources that are self-documenting.

**Comment by BlodyxCZ** [comment]
I like where you are going with your ideas. However, I would like to disagree on the modules. From the examples you have given (specificaly 1, 2, 4), I think modules like that are very project specific, and making something that works generally on all projects is beyond single feature module. I am currently making my own [inventory system](https://github.com/BlodyxCZ/SwiftInv-GodotAddonl) and I don't think it really matches your description. I may just be to pessimistic tho. Something like a debug/dev console or SceneManager could work.

**Comment by coderKillo** [comment]
Love both ideas. Maby we can make some modules depending on the game type people want to create.

I was previously developing in unity and there had something like this in the asset store which come close to what I mean:
https://assetstore.unity.com/packages/templates/systems/topdown-engine-89636?srsltid=AfmBOoo1LWiRdT850GsfiGAJoN7ZgbS-qF6tsekItaiV7ibp8lFcb6YH 

It is like a framework which sits on top of the base game engine. Containing a lot of modules to create a specific type of game (in this case top down dungeon scroller but would also cool for platformers, rouge likes and card games).

These models let the developers focus more on game design instead of implementation and programming.

Also they have some really cool design principles which we can use, like each module can be extended by the developers.
